{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ad224a5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import math\n",
    "from Bio import SeqIO\n",
    "from scipy import stats\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad0fe7ba",
   "metadata": {},
   "source": [
    "## Data wrangling"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1047c46c",
   "metadata": {},
   "source": [
    "Use tracy program to align and get base call intensities for Sanger reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1f428716",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import os\n",
    "import json\n",
    "import re\n",
    "\n",
    "class trace():\n",
    "    \"\"\"Class which represents individual Sanger sequence trace (ab1 files).\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, ab1_path, expected_conversion, ref, output_dir, insert, adar, rnaseh, sample_id):\n",
    "        self.ab1_path = Path(ab1_path)\n",
    "        self.ref = Path(ref)\n",
    "        self.expected_conversion = expected_conversion\n",
    "        self.output_dir = Path(output_dir)\n",
    "        self.insert = insert\n",
    "        self.adar = adar\n",
    "        self.rnaseh = rnaseh\n",
    "        self.sample_id = sample_id\n",
    "    \n",
    "    @property\n",
    "    def basecall_file(self):\n",
    "        return f'{self.output_dir.joinpath(self.ab1_path.stem)}.tracy.basecalls.tsv'\n",
    "    \n",
    "    @property\n",
    "    def align_file(self):\n",
    "        return f'{self.output_dir.joinpath(self.ab1_path.stem)}.tracy.align.json'\n",
    "    \n",
    "\n",
    "    def _tracy_basecall(self):\n",
    "        \"\"\"Use Tracy to generate basecalls from the provided abi file. Since\n",
    "        we are lazy for now assume that Tracy is globally available.\n",
    "\n",
    "        After the basecalls are saved to a tsv file read and return the results as\n",
    "        a pandas dataframe.\n",
    "        \"\"\"\n",
    "\n",
    "        cmd = f\"tracy basecall -f tsv -o {str(self.basecall_file)} {str(self.ab1_path)}\"\n",
    "        os.system(cmd)\n",
    "\n",
    "        return pd.read_csv(self.basecall_file, sep='\\t')\n",
    "    \n",
    "    def _tracy_align(self, basecalls):\n",
    "        \n",
    "        cmd = f'tracy align -l 0 -r 0 -f align -g {str(self.ref)} -o {self.align_file} {str(self.ab1_path)}'\n",
    "        os.system(cmd)\n",
    "        \n",
    "        with open(str(self.align_file)) as handle:\n",
    "            align_text = handle.read()\n",
    "        \n",
    "        # Use regex to pull out alignment text\n",
    "        \n",
    "        ref = ''.join(re.findall('Ref\\s+\\d+\\s(.+)', align_text))\n",
    "        alt = ''.join(re.findall('Alt\\s+\\d+\\s(.+)', align_text))\n",
    "        \n",
    "        # To assign basecalls from Tracy basecall file to a reference nucleotide\n",
    "        # start by assuming everything is a gap. Then if we encounter a gap in the\n",
    "        # trace (alt) alignment skip one basepair in the reference. Without doing\n",
    "        # this gaps in the alt alignment would cause the alignment to be off by\n",
    "        # the total length of gaps compared to the basecalls\n",
    "        \n",
    "        # Tracy sort of put all this information together in its json output but the\n",
    "        # alignments there are not the same as the basecalls within the json file\n",
    "        # or from the output of basecall program so I am not sure what is going\n",
    "        # on there so we have to put them together manually\n",
    "        \n",
    "        nuc_adds = ['-'] * len(basecalls)\n",
    "\n",
    "        positions_added = 0\n",
    "\n",
    "        for r, a in zip(ref, alt):\n",
    "            if a == '-':\n",
    "                continue\n",
    "            else:\n",
    "                nuc_adds[positions_added] = r\n",
    "            positions_added += 1\n",
    "        \n",
    "        basecalls['ref_nuc'] = nuc_adds\n",
    "        \n",
    "        return basecalls\n",
    "    \n",
    "\n",
    "    def analyize(self):\n",
    "        \"\"\"This method is really the core functionality of the script. The idea being we will use\n",
    "        the relative signal that is included as part of the Tracy basecalls in order to locate\n",
    "        bases where editing may have occurred in some population of the molecules present in the\n",
    "        Sanger sequencing reaction. \n",
    "        \"\"\"\n",
    "\n",
    "        # Call bases using Tracy\n",
    "        basecalls = self._tracy_basecall()\n",
    "\n",
    "        # ab1 files include 10 values per base. Tracy retains those values which is nice but to\n",
    "        # simplify we will throw out rows that are not assigned to a nucleotide position. The\n",
    "        # sites that are retained should be the peak crests\n",
    "\n",
    "        bc_trim = basecalls.loc[~basecalls['basenum'].isna()]\n",
    "\n",
    "        assert len(bc_trim) > 0\n",
    "        \n",
    "        \n",
    "        bc_trim_ref = self._tracy_align(bc_trim)\n",
    "        \n",
    "        \n",
    "        bc_trim_ref['conversion_ratio'] = bc_trim_ref['peakC'] / (bc_trim_ref['peakC'] + bc_trim_ref['peakT'])\n",
    "        bc_trim_ref['sum_signal'] = bc_trim_ref['peakA'] + bc_trim_ref['peakT'] + bc_trim_ref['peakG'] + bc_trim_ref['peakC']        \n",
    "        bc_trim_ref['ADAR'] = self.adar\n",
    "        bc_trim_ref['insert'] = self.insert\n",
    "        bc_trim_ref['rnaseh'] = self.rnaseh\n",
    "        bc_trim_ref['sample_ID'] = self.sample_id\n",
    "\n",
    "\n",
    "        return bc_trim_ref"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3cfe4f1",
   "metadata": {},
   "source": [
    "Read ab1 trace files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "66c4df7f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[PosixPath('../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-14_VRI_1__2023-02-04_F04.ab1'),\n",
       " PosixPath('../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-16_VRI_1__2023-02-04_H04.ab1'),\n",
       " PosixPath('../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-6_VRI_1__2023-02-04_B04.ab1'),\n",
       " PosixPath('../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-13_VRI_1__2023-02-04_E04.ab1'),\n",
       " PosixPath('../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-7_VRI_1__2023-02-04_C04.ab1'),\n",
       " PosixPath('../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-8_VRI_1__2023-02-04_D04.ab1'),\n",
       " PosixPath('../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-15_VRI_1__2023-02-04_G04.ab1'),\n",
       " PosixPath('../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-5_VRI_1__2023-02-04_A04.ab1')]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reads_dir = Path('../sangerSeq/2-3-23/2023-02-03_827248')\n",
    "\n",
    "reads = [p for p in reads_dir.iterdir() if p.suffix == '.ab1']\n",
    "\n",
    "reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1f9dcc74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ab1_path</th>\n",
       "      <th>sample_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            ab1_path  sample_ID\n",
       "0  ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...         14\n",
       "1  ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...         16\n",
       "2  ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...          6\n",
       "3  ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...         13\n",
       "4  ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...          7\n",
       "5  ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...          8\n",
       "6  ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...         15\n",
       "7  ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...          5"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reads_list = []\n",
    "\n",
    "for r in reads:\n",
    "    reads_list.append(\n",
    "    {\n",
    "        'ab1_path': str(r),\n",
    "        'sample_ID': int(r.stem.split('_VRI')[0].split('-')[-1])\n",
    "    }\n",
    "    \n",
    "    )\n",
    "reads_df = pd.DataFrame(reads_list)\n",
    "reads_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1077abf6",
   "metadata": {},
   "source": [
    "Reads template files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b886ceba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[PosixPath('../templates/T7_init_VR-26.in.fa'),\n",
       " PosixPath('../templates/T7_init_VR-22.in.fa'),\n",
       " PosixPath('../templates/T7_init_VR-23.in.fa')]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "templates = [p for p in Path('../templates').iterdir() if p.suffix == '.fa']\n",
    "\n",
    "templates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "60148416",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>path</th>\n",
       "      <th>insert</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T7_init_VR-26.in</td>\n",
       "      <td>../templates/T7_init_VR-26.in.fa</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T7_init_VR-22.in</td>\n",
       "      <td>../templates/T7_init_VR-22.in.fa</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T7_init_VR-23.in</td>\n",
       "      <td>../templates/T7_init_VR-23.in.fa</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               name                              path  insert\n",
       "0  T7_init_VR-26.in  ../templates/T7_init_VR-26.in.fa      26\n",
       "1  T7_init_VR-22.in  ../templates/T7_init_VR-22.in.fa      22\n",
       "2  T7_init_VR-23.in  ../templates/T7_init_VR-23.in.fa      23"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temps = []\n",
    "\n",
    "for path in templates:\n",
    "    temps.append({\n",
    "        'name': path.stem,\n",
    "        'path': str(path),\n",
    "        'insert': int(path.stem.split('VR-')[-1].split('.')[0])\n",
    "        \n",
    "    })\n",
    "template_df = pd.DataFrame(temps)\n",
    "template_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c061b043",
   "metadata": {},
   "source": [
    "Read sample assignments into table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1a454155",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = pd.read_csv('../sampleTables/ADAR-IVT-samples.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7716e617",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>tube_label</th>\n",
       "      <th>insert</th>\n",
       "      <th>txn</th>\n",
       "      <th>RnaseH</th>\n",
       "      <th>ADAR_species</th>\n",
       "      <th>IVT_date</th>\n",
       "      <th>ADAR_treatment_date</th>\n",
       "      <th>fwd_primer</th>\n",
       "      <th>rev_primer</th>\n",
       "      <th>fwd_primer_seq</th>\n",
       "      <th>rev_primer_seq</th>\n",
       "      <th>pcr_date</th>\n",
       "      <th>num_cycles</th>\n",
       "      <th>anneal_temp</th>\n",
       "      <th>ext_time</th>\n",
       "      <th>pcr_rxn_volume</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>23 ADAR2 -RH A-</td>\n",
       "      <td>23</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>PacBioFwdV2-0</td>\n",
       "      <td>PacBioRevV2-0</td>\n",
       "      <td>GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...</td>\n",
       "      <td>ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...</td>\n",
       "      <td>2-1-23</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>90</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>23 ADAR2 +RH A-</td>\n",
       "      <td>23</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>PacBioFwdV2-0</td>\n",
       "      <td>PacBioRevV2-0</td>\n",
       "      <td>GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...</td>\n",
       "      <td>ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...</td>\n",
       "      <td>2-1-23</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>90</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>23 ADAR1 -RH A-</td>\n",
       "      <td>23</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>PacBioFwdV2-0</td>\n",
       "      <td>PacBioRevV2-0</td>\n",
       "      <td>GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...</td>\n",
       "      <td>ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...</td>\n",
       "      <td>2-1-23</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>90</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>23 ADAR1 +RH A-</td>\n",
       "      <td>23</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>PacBioFwdV2-0</td>\n",
       "      <td>PacBioRevV2-0</td>\n",
       "      <td>GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...</td>\n",
       "      <td>ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...</td>\n",
       "      <td>2-1-23</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>90</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>23 ADAR2 -RH A+</td>\n",
       "      <td>23</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>PacBioFwdV2-0</td>\n",
       "      <td>PacBioRevV2-0</td>\n",
       "      <td>GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...</td>\n",
       "      <td>ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...</td>\n",
       "      <td>2-1-23</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>90</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_ID       tube_label  insert txn RnaseH ADAR_species IVT_date  \\\n",
       "0          1  23 ADAR2 -RH A-      23   +      -            -  1-31-23   \n",
       "1          2  23 ADAR2 +RH A-      23   +      +            -  1-31-23   \n",
       "2          3  23 ADAR1 -RH A-      23   +      -            -  1-31-23   \n",
       "3          4  23 ADAR1 +RH A-      23   +      +            -  1-31-23   \n",
       "4          5  23 ADAR2 -RH A+      23   +      -            2  1-31-23   \n",
       "\n",
       "  ADAR_treatment_date     fwd_primer     rev_primer  \\\n",
       "0             1-31-23  PacBioFwdV2-0  PacBioRevV2-0   \n",
       "1             1-31-23  PacBioFwdV2-0  PacBioRevV2-0   \n",
       "2             1-31-23  PacBioFwdV2-0  PacBioRevV2-0   \n",
       "3             1-31-23  PacBioFwdV2-0  PacBioRevV2-0   \n",
       "4             1-31-23  PacBioFwdV2-0  PacBioRevV2-0   \n",
       "\n",
       "                                      fwd_primer_seq  \\\n",
       "0  GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...   \n",
       "1  GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...   \n",
       "2  GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...   \n",
       "3  GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...   \n",
       "4  GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...   \n",
       "\n",
       "                                      rev_primer_seq pcr_date  num_cycles  \\\n",
       "0  ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...   2-1-23          30   \n",
       "1  ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...   2-1-23          30   \n",
       "2  ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...   2-1-23          30   \n",
       "3  ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...   2-1-23          30   \n",
       "4  ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...   2-1-23          30   \n",
       "\n",
       "   anneal_temp  ext_time  pcr_rxn_volume  \n",
       "0           72        90              20  \n",
       "1           72        90              20  \n",
       "2           72        90              20  \n",
       "3           72        90              20  \n",
       "4           72        90              20  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cf32df27",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = samples.merge(template_df, on='insert')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d2bd54f",
   "metadata": {},
   "source": [
    "Turn samples dataframe into trace objects. First select samples that were sequenced"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e6379798",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ab1_path</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>tube_label</th>\n",
       "      <th>insert</th>\n",
       "      <th>txn</th>\n",
       "      <th>RnaseH</th>\n",
       "      <th>ADAR_species</th>\n",
       "      <th>IVT_date</th>\n",
       "      <th>ADAR_treatment_date</th>\n",
       "      <th>fwd_primer</th>\n",
       "      <th>rev_primer</th>\n",
       "      <th>fwd_primer_seq</th>\n",
       "      <th>rev_primer_seq</th>\n",
       "      <th>pcr_date</th>\n",
       "      <th>num_cycles</th>\n",
       "      <th>anneal_temp</th>\n",
       "      <th>ext_time</th>\n",
       "      <th>pcr_rxn_volume</th>\n",
       "      <th>name</th>\n",
       "      <th>path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...</td>\n",
       "      <td>14</td>\n",
       "      <td>22 ADAR2 +RH A+</td>\n",
       "      <td>22</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>PacBioFwdV2-0</td>\n",
       "      <td>PacBioRevV2-0</td>\n",
       "      <td>GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...</td>\n",
       "      <td>ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...</td>\n",
       "      <td>2-1-23</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>90</td>\n",
       "      <td>20</td>\n",
       "      <td>T7_init_VR-22.in</td>\n",
       "      <td>../templates/T7_init_VR-22.in.fa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...</td>\n",
       "      <td>16</td>\n",
       "      <td>22 ADAR1 -RH A+</td>\n",
       "      <td>22</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>PacBioFwdV2-0</td>\n",
       "      <td>PacBioRevV2-0</td>\n",
       "      <td>GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...</td>\n",
       "      <td>ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...</td>\n",
       "      <td>2-1-23</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>90</td>\n",
       "      <td>20</td>\n",
       "      <td>T7_init_VR-22.in</td>\n",
       "      <td>../templates/T7_init_VR-22.in.fa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...</td>\n",
       "      <td>6</td>\n",
       "      <td>23 ADAR2 +RH A+</td>\n",
       "      <td>23</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>PacBioFwdV2-0</td>\n",
       "      <td>PacBioRevV2-0</td>\n",
       "      <td>GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...</td>\n",
       "      <td>ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...</td>\n",
       "      <td>2-1-23</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>90</td>\n",
       "      <td>20</td>\n",
       "      <td>T7_init_VR-23.in</td>\n",
       "      <td>../templates/T7_init_VR-23.in.fa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...</td>\n",
       "      <td>13</td>\n",
       "      <td>22 ADAR2 -RH A+</td>\n",
       "      <td>22</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>PacBioFwdV2-0</td>\n",
       "      <td>PacBioRevV2-0</td>\n",
       "      <td>GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...</td>\n",
       "      <td>ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...</td>\n",
       "      <td>2-1-23</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>90</td>\n",
       "      <td>20</td>\n",
       "      <td>T7_init_VR-22.in</td>\n",
       "      <td>../templates/T7_init_VR-22.in.fa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...</td>\n",
       "      <td>7</td>\n",
       "      <td>23 ADAR1 +RH A+</td>\n",
       "      <td>23</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>1-31-23</td>\n",
       "      <td>PacBioFwdV2-0</td>\n",
       "      <td>PacBioRevV2-0</td>\n",
       "      <td>GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...</td>\n",
       "      <td>ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...</td>\n",
       "      <td>2-1-23</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>90</td>\n",
       "      <td>20</td>\n",
       "      <td>T7_init_VR-23.in</td>\n",
       "      <td>../templates/T7_init_VR-23.in.fa</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            ab1_path  sample_ID  \\\n",
       "0  ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...         14   \n",
       "1  ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...         16   \n",
       "2  ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...          6   \n",
       "3  ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...         13   \n",
       "4  ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT...          7   \n",
       "\n",
       "        tube_label  insert txn RnaseH ADAR_species IVT_date  \\\n",
       "0  22 ADAR2 +RH A+      22   +      +            2  1-31-23   \n",
       "1  22 ADAR1 -RH A+      22   +      -            1  1-31-23   \n",
       "2  23 ADAR2 +RH A+      23   +      +            2  1-31-23   \n",
       "3  22 ADAR2 -RH A+      22   +      -            2  1-31-23   \n",
       "4  23 ADAR1 +RH A+      23   +      +            1  1-31-23   \n",
       "\n",
       "  ADAR_treatment_date     fwd_primer     rev_primer  \\\n",
       "0             1-31-23  PacBioFwdV2-0  PacBioRevV2-0   \n",
       "1             1-31-23  PacBioFwdV2-0  PacBioRevV2-0   \n",
       "2             1-31-23  PacBioFwdV2-0  PacBioRevV2-0   \n",
       "3             1-31-23  PacBioFwdV2-0  PacBioRevV2-0   \n",
       "4             1-31-23  PacBioFwdV2-0  PacBioRevV2-0   \n",
       "\n",
       "                                      fwd_primer_seq  \\\n",
       "0  GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...   \n",
       "1  GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...   \n",
       "2  GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...   \n",
       "3  GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...   \n",
       "4  GGCAAACGCGCCATTCGTGTCGGTTAGAAGCACTAAATCGGAACCC...   \n",
       "\n",
       "                                      rev_primer_seq pcr_date  num_cycles  \\\n",
       "0  ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...   2-1-23          30   \n",
       "1  ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...   2-1-23          30   \n",
       "2  ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...   2-1-23          30   \n",
       "3  ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...   2-1-23          30   \n",
       "4  ACATTCGCGCACGTGGTGCACGGTTAGAGGCGAACTACTTACTCTA...   2-1-23          30   \n",
       "\n",
       "   anneal_temp  ext_time  pcr_rxn_volume              name  \\\n",
       "0           72        90              20  T7_init_VR-22.in   \n",
       "1           72        90              20  T7_init_VR-22.in   \n",
       "2           72        90              20  T7_init_VR-23.in   \n",
       "3           72        90              20  T7_init_VR-22.in   \n",
       "4           72        90              20  T7_init_VR-23.in   \n",
       "\n",
       "                               path  \n",
       "0  ../templates/T7_init_VR-22.in.fa  \n",
       "1  ../templates/T7_init_VR-22.in.fa  \n",
       "2  ../templates/T7_init_VR-23.in.fa  \n",
       "3  ../templates/T7_init_VR-22.in.fa  \n",
       "4  ../templates/T7_init_VR-23.in.fa  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exp_1_df = reads_df.merge(samples, on='sample_ID')\n",
    "exp_1_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0b7e561c",
   "metadata": {},
   "outputs": [],
   "source": [
    "Path('sanger-1').mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "71d901a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<__main__.trace at 0x7f9c6669e0b0>,\n",
       " <__main__.trace at 0x7f9c6669f340>,\n",
       " <__main__.trace at 0x7f9c6669ea70>,\n",
       " <__main__.trace at 0x7f9c6669f3a0>,\n",
       " <__main__.trace at 0x7f9c6669f460>,\n",
       " <__main__.trace at 0x7f9c6669f6d0>,\n",
       " <__main__.trace at 0x7f9c6669ee60>,\n",
       " <__main__.trace at 0x7f9c6669ebf0>]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "traces = []\n",
    "\n",
    "for i, row in exp_1_df.iterrows():\n",
    "    traces.append(\n",
    "        trace(\n",
    "            row.ab1_path, ('T', 'C'), \n",
    "            row.path, \n",
    "            'sanger-1', \n",
    "            row.insert, \n",
    "            row.ADAR_species, \n",
    "            row.RnaseH,\n",
    "            row.sample_ID)\n",
    "    )\n",
    "traces"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "dbed24d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2023-Feb-06 13:10:48] tracy basecall -f tsv -o sanger-1/ADAR-IVT-14_VRI_1__2023-02-04_F04.tracy.basecalls.tsv ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-14_VRI_1__2023-02-04_F04.ab1 \n",
      "[2023-Feb-06 13:10:48] Done.\n",
      "[2023-Feb-06 13:10:48] tracy align -l 0 -r 0 -f align -g ../templates/T7_init_VR-22.in.fa -o sanger-1/ADAR-IVT-14_VRI_1__2023-02-04_F04.tracy.align.json ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-14_VRI_1__2023-02-04_F04.ab1 \n",
      "[2023-Feb-06 13:10:48] Load ab1 file\n",
      "[2023-Feb-06 13:10:48] Load FASTA reference\n",
      "[2023-Feb-06 13:10:48] Find reference match\n",
      "[2023-Feb-06 13:10:48] Alignment\n",
      "[2023-Feb-06 13:10:48] Output\n",
      "[2023-Feb-06 13:10:48] Done.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_20845/3961057931.py:77: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  basecalls['ref_nuc'] = nuc_adds\n",
      "/tmp/ipykernel_20845/3961057931.py:104: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['conversion_ratio'] = bc_trim_ref['peakC'] / (bc_trim_ref['peakC'] + bc_trim_ref['peakT'])\n",
      "/tmp/ipykernel_20845/3961057931.py:105: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['sum_signal'] = bc_trim_ref['peakA'] + bc_trim_ref['peakT'] + bc_trim_ref['peakG'] + bc_trim_ref['peakC']\n",
      "/tmp/ipykernel_20845/3961057931.py:106: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['ADAR'] = self.adar\n",
      "/tmp/ipykernel_20845/3961057931.py:107: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['insert'] = self.insert\n",
      "/tmp/ipykernel_20845/3961057931.py:108: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['rnaseh'] = self.rnaseh\n",
      "/tmp/ipykernel_20845/3961057931.py:109: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['sample_ID'] = self.sample_id\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2023-Feb-06 13:10:48] tracy basecall -f tsv -o sanger-1/ADAR-IVT-16_VRI_1__2023-02-04_H04.tracy.basecalls.tsv ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-16_VRI_1__2023-02-04_H04.ab1 \n",
      "[2023-Feb-06 13:10:48] Done.\n",
      "[2023-Feb-06 13:10:48] tracy align -l 0 -r 0 -f align -g ../templates/T7_init_VR-22.in.fa -o sanger-1/ADAR-IVT-16_VRI_1__2023-02-04_H04.tracy.align.json ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-16_VRI_1__2023-02-04_H04.ab1 \n",
      "[2023-Feb-06 13:10:48] Load ab1 file\n",
      "[2023-Feb-06 13:10:48] Load FASTA reference\n",
      "[2023-Feb-06 13:10:48] Find reference match\n",
      "[2023-Feb-06 13:10:48] Alignment\n",
      "[2023-Feb-06 13:10:48] Output\n",
      "[2023-Feb-06 13:10:48] Done.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_20845/3961057931.py:77: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  basecalls['ref_nuc'] = nuc_adds\n",
      "/tmp/ipykernel_20845/3961057931.py:104: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['conversion_ratio'] = bc_trim_ref['peakC'] / (bc_trim_ref['peakC'] + bc_trim_ref['peakT'])\n",
      "/tmp/ipykernel_20845/3961057931.py:105: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['sum_signal'] = bc_trim_ref['peakA'] + bc_trim_ref['peakT'] + bc_trim_ref['peakG'] + bc_trim_ref['peakC']\n",
      "/tmp/ipykernel_20845/3961057931.py:106: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['ADAR'] = self.adar\n",
      "/tmp/ipykernel_20845/3961057931.py:107: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['insert'] = self.insert\n",
      "/tmp/ipykernel_20845/3961057931.py:108: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['rnaseh'] = self.rnaseh\n",
      "/tmp/ipykernel_20845/3961057931.py:109: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['sample_ID'] = self.sample_id\n",
      "/tmp/ipykernel_20845/3961057931.py:77: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  basecalls['ref_nuc'] = nuc_adds\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2023-Feb-06 13:10:48] tracy basecall -f tsv -o sanger-1/ADAR-IVT-6_VRI_1__2023-02-04_B04.tracy.basecalls.tsv ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-6_VRI_1__2023-02-04_B04.ab1 \n",
      "[2023-Feb-06 13:10:48] Done.\n",
      "[2023-Feb-06 13:10:48] tracy align -l 0 -r 0 -f align -g ../templates/T7_init_VR-23.in.fa -o sanger-1/ADAR-IVT-6_VRI_1__2023-02-04_B04.tracy.align.json ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-6_VRI_1__2023-02-04_B04.ab1 \n",
      "[2023-Feb-06 13:10:48] Load ab1 file\n",
      "[2023-Feb-06 13:10:48] Load FASTA reference\n",
      "[2023-Feb-06 13:10:48] Find reference match\n",
      "[2023-Feb-06 13:10:48] Alignment\n",
      "[2023-Feb-06 13:10:48] Output\n",
      "[2023-Feb-06 13:10:48] Done.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_20845/3961057931.py:104: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['conversion_ratio'] = bc_trim_ref['peakC'] / (bc_trim_ref['peakC'] + bc_trim_ref['peakT'])\n",
      "/tmp/ipykernel_20845/3961057931.py:105: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['sum_signal'] = bc_trim_ref['peakA'] + bc_trim_ref['peakT'] + bc_trim_ref['peakG'] + bc_trim_ref['peakC']\n",
      "/tmp/ipykernel_20845/3961057931.py:106: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['ADAR'] = self.adar\n",
      "/tmp/ipykernel_20845/3961057931.py:107: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['insert'] = self.insert\n",
      "/tmp/ipykernel_20845/3961057931.py:108: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['rnaseh'] = self.rnaseh\n",
      "/tmp/ipykernel_20845/3961057931.py:109: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['sample_ID'] = self.sample_id\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2023-Feb-06 13:10:48] tracy basecall -f tsv -o sanger-1/ADAR-IVT-13_VRI_1__2023-02-04_E04.tracy.basecalls.tsv ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-13_VRI_1__2023-02-04_E04.ab1 \n",
      "[2023-Feb-06 13:10:48] Done.\n",
      "[2023-Feb-06 13:10:49] tracy align -l 0 -r 0 -f align -g ../templates/T7_init_VR-22.in.fa -o sanger-1/ADAR-IVT-13_VRI_1__2023-02-04_E04.tracy.align.json ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-13_VRI_1__2023-02-04_E04.ab1 \n",
      "[2023-Feb-06 13:10:49] Load ab1 file\n",
      "[2023-Feb-06 13:10:49] Load FASTA reference\n",
      "[2023-Feb-06 13:10:49] Find reference match\n",
      "[2023-Feb-06 13:10:49] Alignment\n",
      "[2023-Feb-06 13:10:49] Output\n",
      "[2023-Feb-06 13:10:49] Done.\n",
      "[2023-Feb-06 13:10:49] tracy basecall -f tsv -o sanger-1/ADAR-IVT-7_VRI_1__2023-02-04_C04.tracy.basecalls.tsv ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-7_VRI_1__2023-02-04_C04.ab1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_20845/3961057931.py:77: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  basecalls['ref_nuc'] = nuc_adds\n",
      "/tmp/ipykernel_20845/3961057931.py:104: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['conversion_ratio'] = bc_trim_ref['peakC'] / (bc_trim_ref['peakC'] + bc_trim_ref['peakT'])\n",
      "/tmp/ipykernel_20845/3961057931.py:105: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['sum_signal'] = bc_trim_ref['peakA'] + bc_trim_ref['peakT'] + bc_trim_ref['peakG'] + bc_trim_ref['peakC']\n",
      "/tmp/ipykernel_20845/3961057931.py:106: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['ADAR'] = self.adar\n",
      "/tmp/ipykernel_20845/3961057931.py:107: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['insert'] = self.insert\n",
      "/tmp/ipykernel_20845/3961057931.py:108: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['rnaseh'] = self.rnaseh\n",
      "/tmp/ipykernel_20845/3961057931.py:109: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['sample_ID'] = self.sample_id\n",
      "/tmp/ipykernel_20845/3961057931.py:77: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  basecalls['ref_nuc'] = nuc_adds\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2023-Feb-06 13:10:49] Done.\n",
      "[2023-Feb-06 13:10:49] tracy align -l 0 -r 0 -f align -g ../templates/T7_init_VR-23.in.fa -o sanger-1/ADAR-IVT-7_VRI_1__2023-02-04_C04.tracy.align.json ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-7_VRI_1__2023-02-04_C04.ab1 \n",
      "[2023-Feb-06 13:10:49] Load ab1 file\n",
      "[2023-Feb-06 13:10:49] Load FASTA reference\n",
      "[2023-Feb-06 13:10:49] Find reference match\n",
      "[2023-Feb-06 13:10:49] Alignment\n",
      "[2023-Feb-06 13:10:49] Output\n",
      "[2023-Feb-06 13:10:49] Done.\n",
      "[2023-Feb-06 13:10:49] tracy basecall -f tsv -o sanger-1/ADAR-IVT-8_VRI_1__2023-02-04_D04.tracy.basecalls.tsv ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-8_VRI_1__2023-02-04_D04.ab1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_20845/3961057931.py:104: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['conversion_ratio'] = bc_trim_ref['peakC'] / (bc_trim_ref['peakC'] + bc_trim_ref['peakT'])\n",
      "/tmp/ipykernel_20845/3961057931.py:105: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['sum_signal'] = bc_trim_ref['peakA'] + bc_trim_ref['peakT'] + bc_trim_ref['peakG'] + bc_trim_ref['peakC']\n",
      "/tmp/ipykernel_20845/3961057931.py:106: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['ADAR'] = self.adar\n",
      "/tmp/ipykernel_20845/3961057931.py:107: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['insert'] = self.insert\n",
      "/tmp/ipykernel_20845/3961057931.py:108: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['rnaseh'] = self.rnaseh\n",
      "/tmp/ipykernel_20845/3961057931.py:109: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['sample_ID'] = self.sample_id\n",
      "/tmp/ipykernel_20845/3961057931.py:77: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  basecalls['ref_nuc'] = nuc_adds\n",
      "/tmp/ipykernel_20845/3961057931.py:104: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['conversion_ratio'] = bc_trim_ref['peakC'] / (bc_trim_ref['peakC'] + bc_trim_ref['peakT'])\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2023-Feb-06 13:10:49] Done.\n",
      "[2023-Feb-06 13:10:49] tracy align -l 0 -r 0 -f align -g ../templates/T7_init_VR-23.in.fa -o sanger-1/ADAR-IVT-8_VRI_1__2023-02-04_D04.tracy.align.json ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-8_VRI_1__2023-02-04_D04.ab1 \n",
      "[2023-Feb-06 13:10:49] Load ab1 file\n",
      "[2023-Feb-06 13:10:49] Load FASTA reference\n",
      "[2023-Feb-06 13:10:49] Find reference match\n",
      "[2023-Feb-06 13:10:49] Alignment\n",
      "[2023-Feb-06 13:10:49] Output\n",
      "[2023-Feb-06 13:10:49] Done.\n",
      "[2023-Feb-06 13:10:49] tracy basecall -f tsv -o sanger-1/ADAR-IVT-15_VRI_1__2023-02-04_G04.tracy.basecalls.tsv ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-15_VRI_1__2023-02-04_G04.ab1 \n",
      "[2023-Feb-06 13:10:49] Done.\n",
      "[2023-Feb-06 13:10:49] tracy align -l 0 -r 0 -f align -g ../templates/T7_init_VR-22.in.fa -o sanger-1/ADAR-IVT-15_VRI_1__2023-02-04_G04.tracy.align.json ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-15_VRI_1__2023-02-04_G04.ab1 \n",
      "[2023-Feb-06 13:10:49] Load ab1 file\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_20845/3961057931.py:105: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['sum_signal'] = bc_trim_ref['peakA'] + bc_trim_ref['peakT'] + bc_trim_ref['peakG'] + bc_trim_ref['peakC']\n",
      "/tmp/ipykernel_20845/3961057931.py:106: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['ADAR'] = self.adar\n",
      "/tmp/ipykernel_20845/3961057931.py:107: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['insert'] = self.insert\n",
      "/tmp/ipykernel_20845/3961057931.py:108: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['rnaseh'] = self.rnaseh\n",
      "/tmp/ipykernel_20845/3961057931.py:109: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['sample_ID'] = self.sample_id\n",
      "/tmp/ipykernel_20845/3961057931.py:77: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  basecalls['ref_nuc'] = nuc_adds\n",
      "/tmp/ipykernel_20845/3961057931.py:104: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['conversion_ratio'] = bc_trim_ref['peakC'] / (bc_trim_ref['peakC'] + bc_trim_ref['peakT'])\n",
      "/tmp/ipykernel_20845/3961057931.py:105: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['sum_signal'] = bc_trim_ref['peakA'] + bc_trim_ref['peakT'] + bc_trim_ref['peakG'] + bc_trim_ref['peakC']\n",
      "/tmp/ipykernel_20845/3961057931.py:106: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['ADAR'] = self.adar\n",
      "/tmp/ipykernel_20845/3961057931.py:107: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['insert'] = self.insert\n",
      "/tmp/ipykernel_20845/3961057931.py:108: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['rnaseh'] = self.rnaseh\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2023-Feb-06 13:10:49] Load FASTA reference\n",
      "[2023-Feb-06 13:10:49] Find reference match\n",
      "[2023-Feb-06 13:10:49] Alignment\n",
      "[2023-Feb-06 13:10:49] Output\n",
      "[2023-Feb-06 13:10:49] Done.\n",
      "[2023-Feb-06 13:10:49] tracy basecall -f tsv -o sanger-1/ADAR-IVT-5_VRI_1__2023-02-04_A04.tracy.basecalls.tsv ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-5_VRI_1__2023-02-04_A04.ab1 \n",
      "[2023-Feb-06 13:10:49] Done.\n",
      "[2023-Feb-06 13:10:49] tracy align -l 0 -r 0 -f align -g ../templates/T7_init_VR-23.in.fa -o sanger-1/ADAR-IVT-5_VRI_1__2023-02-04_A04.tracy.align.json ../sangerSeq/2-3-23/2023-02-03_827248/ADAR-IVT-5_VRI_1__2023-02-04_A04.ab1 \n",
      "[2023-Feb-06 13:10:49] Load ab1 file\n",
      "[2023-Feb-06 13:10:49] Load FASTA reference\n",
      "[2023-Feb-06 13:10:49] Find reference match\n",
      "[2023-Feb-06 13:10:49] Alignment\n",
      "[2023-Feb-06 13:10:49] Output\n",
      "[2023-Feb-06 13:10:49] Done.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_20845/3961057931.py:109: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['sample_ID'] = self.sample_id\n",
      "/tmp/ipykernel_20845/3961057931.py:77: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  basecalls['ref_nuc'] = nuc_adds\n",
      "/tmp/ipykernel_20845/3961057931.py:104: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['conversion_ratio'] = bc_trim_ref['peakC'] / (bc_trim_ref['peakC'] + bc_trim_ref['peakT'])\n",
      "/tmp/ipykernel_20845/3961057931.py:105: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['sum_signal'] = bc_trim_ref['peakA'] + bc_trim_ref['peakT'] + bc_trim_ref['peakG'] + bc_trim_ref['peakC']\n",
      "/tmp/ipykernel_20845/3961057931.py:106: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['ADAR'] = self.adar\n",
      "/tmp/ipykernel_20845/3961057931.py:107: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['insert'] = self.insert\n",
      "/tmp/ipykernel_20845/3961057931.py:108: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['rnaseh'] = self.rnaseh\n",
      "/tmp/ipykernel_20845/3961057931.py:109: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bc_trim_ref['sample_ID'] = self.sample_id\n"
     ]
    }
   ],
   "source": [
    "trace_data = []\n",
    "\n",
    "for each_trace in traces:\n",
    "    trace_data.append(each_trace.analyize())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac35706c",
   "metadata": {},
   "source": [
    "Remove outlier peaks from the dataframe using IQR method. This removes peaks with very low or very high signal from each read"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "93446e88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def remove_outlier_peaks(df):\n",
    "    return df[(np.abs(stats.zscore(df['sum_signal'])) < 3)]\n",
    "\n",
    "\n",
    "trace_data_prune = []\n",
    "\n",
    "for each_frame in trace_data:\n",
    "    trace_data_prune.append(remove_outlier_peaks(each_frame))\n",
    "\n",
    "len(trace_data_prune)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b08a28a",
   "metadata": {},
   "source": [
    "Combine dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "51a43a32",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_trace_df = pd.concat(trace_data_prune)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "595aaabf",
   "metadata": {},
   "source": [
    "## Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9219c7aa",
   "metadata": {},
   "source": [
    "### Determine the C / C + T value at locations that should be T according to the aligned reference sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b7126592",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_trace_df_T = all_trace_df.loc[all_trace_df['ref_nuc'] == 'T']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f3890c03",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>peakA</th>\n",
       "      <th>peakC</th>\n",
       "      <th>peakG</th>\n",
       "      <th>peakT</th>\n",
       "      <th>basenum</th>\n",
       "      <th>primary</th>\n",
       "      <th>secondary</th>\n",
       "      <th>consensus</th>\n",
       "      <th>qual</th>\n",
       "      <th>trim</th>\n",
       "      <th>ref_nuc</th>\n",
       "      <th>conversion_ratio</th>\n",
       "      <th>sum_signal</th>\n",
       "      <th>ADAR</th>\n",
       "      <th>insert</th>\n",
       "      <th>rnaseh</th>\n",
       "      <th>sample_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2742</th>\n",
       "      <td>2743</td>\n",
       "      <td>49</td>\n",
       "      <td>14</td>\n",
       "      <td>53</td>\n",
       "      <td>845</td>\n",
       "      <td>233.0</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>49.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>T</td>\n",
       "      <td>0.016298</td>\n",
       "      <td>961</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2754</th>\n",
       "      <td>2755</td>\n",
       "      <td>52</td>\n",
       "      <td>35</td>\n",
       "      <td>30</td>\n",
       "      <td>774</td>\n",
       "      <td>234.0</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>50.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>T</td>\n",
       "      <td>0.043263</td>\n",
       "      <td>891</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2790</th>\n",
       "      <td>2791</td>\n",
       "      <td>66</td>\n",
       "      <td>31</td>\n",
       "      <td>12</td>\n",
       "      <td>454</td>\n",
       "      <td>237.0</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>51.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>T</td>\n",
       "      <td>0.063918</td>\n",
       "      <td>563</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2862</th>\n",
       "      <td>2863</td>\n",
       "      <td>7</td>\n",
       "      <td>97</td>\n",
       "      <td>50</td>\n",
       "      <td>424</td>\n",
       "      <td>243.0</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>34.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>T</td>\n",
       "      <td>0.186180</td>\n",
       "      <td>578</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3076</th>\n",
       "      <td>3077</td>\n",
       "      <td>33</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>849</td>\n",
       "      <td>261.0</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>56.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>T</td>\n",
       "      <td>0.019630</td>\n",
       "      <td>899</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11298</th>\n",
       "      <td>11299</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>21</td>\n",
       "      <td>949.0</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>19.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>37</td>\n",
       "      <td>2</td>\n",
       "      <td>23</td>\n",
       "      <td>-</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11663</th>\n",
       "      <td>11664</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>17</td>\n",
       "      <td>980.0</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>23.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>28</td>\n",
       "      <td>2</td>\n",
       "      <td>23</td>\n",
       "      <td>-</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11812</th>\n",
       "      <td>11813</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>993.0</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>31.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>T</td>\n",
       "      <td>0.032258</td>\n",
       "      <td>33</td>\n",
       "      <td>2</td>\n",
       "      <td>23</td>\n",
       "      <td>-</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11819</th>\n",
       "      <td>11820</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>36</td>\n",
       "      <td>994.0</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>25.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>41</td>\n",
       "      <td>2</td>\n",
       "      <td>23</td>\n",
       "      <td>-</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11856</th>\n",
       "      <td>11857</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>39</td>\n",
       "      <td>997.0</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>32.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>T</td>\n",
       "      <td>0.170213</td>\n",
       "      <td>51</td>\n",
       "      <td>2</td>\n",
       "      <td>23</td>\n",
       "      <td>-</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1319 rows  18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         pos  peakA  peakC  peakG  peakT  basenum primary secondary consensus  \\\n",
       "2742    2743     49     14     53    845    233.0       T         T         T   \n",
       "2754    2755     52     35     30    774    234.0       T         T         T   \n",
       "2790    2791     66     31     12    454    237.0       T         T         T   \n",
       "2862    2863      7     97     50    424    243.0       T         T         T   \n",
       "3076    3077     33     17      0    849    261.0       T         T         T   \n",
       "...      ...    ...    ...    ...    ...      ...     ...       ...       ...   \n",
       "11298  11299      2      0     14     21    949.0       T         G         N   \n",
       "11663  11664      5      0      6     17    980.0       T         G         N   \n",
       "11812  11813      0      1      2     30    993.0       T         T         T   \n",
       "11819  11820      0      0      5     36    994.0       T         T         T   \n",
       "11856  11857      2      8      2     39    997.0       T         T         T   \n",
       "\n",
       "       qual trim ref_nuc  conversion_ratio  sum_signal ADAR  insert rnaseh  \\\n",
       "2742   49.0    Y       T          0.016298         961    2      22      +   \n",
       "2754   50.0    Y       T          0.043263         891    2      22      +   \n",
       "2790   51.0    Y       T          0.063918         563    2      22      +   \n",
       "2862   34.0    Y       T          0.186180         578    2      22      +   \n",
       "3076   56.0    Y       T          0.019630         899    2      22      +   \n",
       "...     ...  ...     ...               ...         ...  ...     ...    ...   \n",
       "11298  19.0    Y       T          0.000000          37    2      23      -   \n",
       "11663  23.0    Y       T          0.000000          28    2      23      -   \n",
       "11812  31.0    Y       T          0.032258          33    2      23      -   \n",
       "11819  25.0    Y       T          0.000000          41    2      23      -   \n",
       "11856  32.0    Y       T          0.170213          51    2      23      -   \n",
       "\n",
       "       sample_ID  \n",
       "2742          14  \n",
       "2754          14  \n",
       "2790          14  \n",
       "2862          14  \n",
       "3076          14  \n",
       "...          ...  \n",
       "11298          5  \n",
       "11663          5  \n",
       "11812          5  \n",
       "11819          5  \n",
       "11856          5  \n",
       "\n",
       "[1319 rows x 18 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_trace_df_T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "cde145e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_trace_df.to_csv('test_all.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f4550ef",
   "metadata": {},
   "source": [
    "Plot this ratio for each condition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f080d7c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "51888fff",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "library(ggplot2)\n",
    "library(ggpubr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b694bc1b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Saving 15 x 5 in image\n",
      "\n",
      "R[write to console]: Error in plot_theme(plot) : object 'plt2' not found\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Error in plot_theme(plot) : object 'plt2' not found\n"
     ]
    },
    {
     "ename": "RInterpreterError",
     "evalue": "Failed to parse and evaluate line \"\\nadar1.df <- subset(all_trace_df_T, ADAR==1)\\nadar2.df <- subset(all_trace_df_T, ADAR==2)\\n\\nadar1.plt <- ggplot(adar1.df, aes(x=basenum, y=conversion_ratio)) + \\n        geom_point(color='black', pch=21, fill='tan', alpha=0.6) + \\n        facet_grid(row=vars(insert), col=vars(rnaseh)) + theme_pubclean() +\\n        labs(y='C / (C+T)', x='Read position', title='ADAR1 Deaimination')\\n\\nadar2.plt <- ggplot(adar2.df, aes(x=basenum, y=conversion_ratio)) + \\n        geom_point(color='black', pch=21, fill='dodgerblue', alpha=0.6) + \\n        facet_grid(row=vars(insert), col=vars(rnaseh)) + theme_pubclean() +\\n        labs(y='C / (C+T)', x='Read position', title='ADAR2 Deaimination')\\n\\nplt.1 <- ggarrange(adar1.plt, adar2.plt)\\nggsave('plots/global_C_to_T_ratio.png', plt2, dpi=300)\\n\".\nR error message: \"Error in plot_theme(plot) : object 'plt2' not found\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRRuntimeError\u001b[0m                             Traceback (most recent call last)",
      "File \u001b[0;32m~/anaconda3/envs/bs/lib/python3.10/site-packages/rpy2/ipython/rmagic.py:280\u001b[0m, in \u001b[0;36mRMagics.eval\u001b[0;34m(self, code)\u001b[0m\n\u001b[1;32m    278\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    279\u001b[0m     \u001b[38;5;66;03m# Need the newline in case the last line in code is a comment.\u001b[39;00m\n\u001b[0;32m--> 280\u001b[0m     value, visible \u001b[38;5;241m=\u001b[39m \u001b[43mro\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mr\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mwithVisible(\u001b[39;49m\u001b[38;5;124;43m{\u001b[39;49m\u001b[38;5;132;43;01m%s\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m})\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m%\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mcode\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    281\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ri\u001b[38;5;241m.\u001b[39membedded\u001b[38;5;241m.\u001b[39mRRuntimeError, \u001b[38;5;167;01mValueError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m exception:\n\u001b[1;32m    282\u001b[0m     \u001b[38;5;66;03m# Otherwise next return seems to have copy of error.\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/bs/lib/python3.10/site-packages/rpy2/robjects/__init__.py:451\u001b[0m, in \u001b[0;36mR.__call__\u001b[0;34m(self, string)\u001b[0m\n\u001b[1;32m    450\u001b[0m p \u001b[38;5;241m=\u001b[39m rinterface\u001b[38;5;241m.\u001b[39mparse(string)\n\u001b[0;32m--> 451\u001b[0m res \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43meval\u001b[49m\u001b[43m(\u001b[49m\u001b[43mp\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    452\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m conversion\u001b[38;5;241m.\u001b[39mrpy2py(res)\n",
      "File \u001b[0;32m~/anaconda3/envs/bs/lib/python3.10/site-packages/rpy2/robjects/functions.py:201\u001b[0m, in \u001b[0;36mSignatureTranslatedFunction.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    200\u001b[0m         kwargs[r_k] \u001b[38;5;241m=\u001b[39m v\n\u001b[0;32m--> 201\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m (\u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mSignatureTranslatedFunction\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    202\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__call__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m)\n",
      "File \u001b[0;32m~/anaconda3/envs/bs/lib/python3.10/site-packages/rpy2/robjects/functions.py:124\u001b[0m, in \u001b[0;36mFunction.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    123\u001b[0m         new_kwargs[k] \u001b[38;5;241m=\u001b[39m conversion\u001b[38;5;241m.\u001b[39mpy2rpy(v)\n\u001b[0;32m--> 124\u001b[0m res \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mFunction\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__call__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mnew_args\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mnew_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    125\u001b[0m res \u001b[38;5;241m=\u001b[39m conversion\u001b[38;5;241m.\u001b[39mrpy2py(res)\n",
      "File \u001b[0;32m~/anaconda3/envs/bs/lib/python3.10/site-packages/rpy2/rinterface_lib/conversion.py:45\u001b[0m, in \u001b[0;36m_cdata_res_to_rinterface.<locals>._\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     44\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_\u001b[39m(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[0;32m---> 45\u001b[0m     cdata \u001b[38;5;241m=\u001b[39m \u001b[43mfunction\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     46\u001b[0m     \u001b[38;5;66;03m# TODO: test cdata is of the expected CType\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/bs/lib/python3.10/site-packages/rpy2/rinterface.py:810\u001b[0m, in \u001b[0;36mSexpClosure.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    809\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m error_occured[\u001b[38;5;241m0\u001b[39m]:\n\u001b[0;32m--> 810\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m embedded\u001b[38;5;241m.\u001b[39mRRuntimeError(_rinterface\u001b[38;5;241m.\u001b[39m_geterrmessage())\n\u001b[1;32m    811\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m res\n",
      "\u001b[0;31mRRuntimeError\u001b[0m: Error in plot_theme(plot) : object 'plt2' not found\n",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mRInterpreterError\u001b[0m                         Traceback (most recent call last)",
      "Input \u001b[0;32mIn [21]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mget_ipython\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_cell_magic\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mR\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m-i all_trace_df_T -w 15 -h 5 --units in -r 400\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43madar1.df <- subset(all_trace_df_T, ADAR==1)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43madar2.df <- subset(all_trace_df_T, ADAR==2)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43madar1.plt <- ggplot(adar1.df, aes(x=basenum, y=conversion_ratio)) + \u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m        geom_point(color=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mblack\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m, pch=21, fill=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mtan\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m, alpha=0.6) + \u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m        facet_grid(row=vars(insert), col=vars(rnaseh)) + theme_pubclean() +\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m        labs(y=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mC / (C+T)\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m, x=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mRead position\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m, title=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mADAR1 Deaimination\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43madar2.plt <- ggplot(adar2.df, aes(x=basenum, y=conversion_ratio)) + \u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m        geom_point(color=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mblack\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m, pch=21, fill=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mdodgerblue\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m, alpha=0.6) + \u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m        facet_grid(row=vars(insert), col=vars(rnaseh)) + theme_pubclean() +\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m        labs(y=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mC / (C+T)\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m, x=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mRead position\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m, title=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mADAR2 Deaimination\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mplt.1 <- ggarrange(adar1.plt, adar2.plt)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mggsave(\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mplots/global_C_to_T_ratio.png\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m, plt2, dpi=300)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/bs/lib/python3.10/site-packages/IPython/core/interactiveshell.py:2357\u001b[0m, in \u001b[0;36mInteractiveShell.run_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2355\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuiltin_trap:\n\u001b[1;32m   2356\u001b[0m     args \u001b[38;5;241m=\u001b[39m (magic_arg_s, cell)\n\u001b[0;32m-> 2357\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2358\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/anaconda3/envs/bs/lib/python3.10/site-packages/rpy2/ipython/rmagic.py:795\u001b[0m, in \u001b[0;36mRMagics.R\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n\u001b[1;32m    793\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m e\u001b[38;5;241m.\u001b[39mstdout\u001b[38;5;241m.\u001b[39mendswith(e\u001b[38;5;241m.\u001b[39merr):\n\u001b[1;32m    794\u001b[0m         \u001b[38;5;28mprint\u001b[39m(e\u001b[38;5;241m.\u001b[39merr)\n\u001b[0;32m--> 795\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\n\u001b[1;32m    796\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    797\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdevice \u001b[38;5;129;01min\u001b[39;00m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpng\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124msvg\u001b[39m\u001b[38;5;124m'\u001b[39m]:\n",
      "File \u001b[0;32m~/anaconda3/envs/bs/lib/python3.10/site-packages/rpy2/ipython/rmagic.py:775\u001b[0m, in \u001b[0;36mRMagics.R\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n\u001b[1;32m    773\u001b[0m         return_output \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m    774\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 775\u001b[0m     text_result, result, visible \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43meval\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcode\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    776\u001b[0m     text_output \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m text_result\n\u001b[1;32m    777\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m visible:\n",
      "File \u001b[0;32m~/anaconda3/envs/bs/lib/python3.10/site-packages/rpy2/ipython/rmagic.py:284\u001b[0m, in \u001b[0;36mRMagics.eval\u001b[0;34m(self, code)\u001b[0m\n\u001b[1;32m    281\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ri\u001b[38;5;241m.\u001b[39membedded\u001b[38;5;241m.\u001b[39mRRuntimeError, \u001b[38;5;167;01mValueError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m exception:\n\u001b[1;32m    282\u001b[0m     \u001b[38;5;66;03m# Otherwise next return seems to have copy of error.\u001b[39;00m\n\u001b[1;32m    283\u001b[0m     warning_or_other_msg \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mflush()\n\u001b[0;32m--> 284\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m RInterpreterError(code, \u001b[38;5;28mstr\u001b[39m(exception),\n\u001b[1;32m    285\u001b[0m                             warning_or_other_msg)\n\u001b[1;32m    286\u001b[0m text_output \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mflush()\n\u001b[1;32m    287\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m text_output, value, visible[\u001b[38;5;241m0\u001b[39m]\n",
      "\u001b[0;31mRInterpreterError\u001b[0m: Failed to parse and evaluate line \"\\nadar1.df <- subset(all_trace_df_T, ADAR==1)\\nadar2.df <- subset(all_trace_df_T, ADAR==2)\\n\\nadar1.plt <- ggplot(adar1.df, aes(x=basenum, y=conversion_ratio)) + \\n        geom_point(color='black', pch=21, fill='tan', alpha=0.6) + \\n        facet_grid(row=vars(insert), col=vars(rnaseh)) + theme_pubclean() +\\n        labs(y='C / (C+T)', x='Read position', title='ADAR1 Deaimination')\\n\\nadar2.plt <- ggplot(adar2.df, aes(x=basenum, y=conversion_ratio)) + \\n        geom_point(color='black', pch=21, fill='dodgerblue', alpha=0.6) + \\n        facet_grid(row=vars(insert), col=vars(rnaseh)) + theme_pubclean() +\\n        labs(y='C / (C+T)', x='Read position', title='ADAR2 Deaimination')\\n\\nplt.1 <- ggarrange(adar1.plt, adar2.plt)\\nggsave('plots/global_C_to_T_ratio.png', plt2, dpi=300)\\n\".\nR error message: \"Error in plot_theme(plot) : object 'plt2' not found\""
     ]
    }
   ],
   "source": [
    "%%R -i all_trace_df_T -w 15 -h 5 --units in -r 400\n",
    "\n",
    "adar1.df <- subset(all_trace_df_T, ADAR==1)\n",
    "adar2.df <- subset(all_trace_df_T, ADAR==2)\n",
    "\n",
    "adar1.plt <- ggplot(adar1.df, aes(x=basenum, y=conversion_ratio)) + \n",
    "        geom_point(color='black', pch=21, fill='tan', alpha=0.6) + \n",
    "        facet_grid(row=vars(insert), col=vars(rnaseh)) + theme_pubclean() +\n",
    "        labs(y='C / (C+T)', x='Read position', title='ADAR1 Deaimination')\n",
    "\n",
    "adar2.plt <- ggplot(adar2.df, aes(x=basenum, y=conversion_ratio)) + \n",
    "        geom_point(color='black', pch=21, fill='dodgerblue', alpha=0.6) + \n",
    "        facet_grid(row=vars(insert), col=vars(rnaseh)) + theme_pubclean() +\n",
    "        labs(y='C / (C+T)', x='Read position', title='ADAR2 Deaimination')\n",
    "\n",
    "plt.1 <- ggarrange(adar1.plt, adar2.plt)\n",
    "ggsave('plots/global_C_to_T_ratio.png', plt2, dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62a53987",
   "metadata": {},
   "source": [
    "## Determine conversion ratios at \"CTA\" sites which should be the best editing canidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84fabcd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def identify_cta_sites(df):\n",
    "    # Determine the row numbers of \"T\" nucleotides in a reference that are part of a CTA site\n",
    "    # and only return those rows\n",
    "    ref_nucs = ''.join(list(df['ref_nuc']))\n",
    "    finds = re.finditer('CTA', ref_nucs)\n",
    "    CTA_T_rows = []\n",
    "    for f in finds:\n",
    "        CTA_T_rows.append(f.span()[0]+1)\n",
    "    return df.iloc[CTA_T_rows]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c459340",
   "metadata": {},
   "outputs": [],
   "source": [
    "CTA_sites = identify_cta_sites(all_trace_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f21c9b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i CTA_sites -w 15 -h 5 --units in -r 400\n",
    "\n",
    "adar1.df.cta <- subset(CTA_sites, ADAR==1)\n",
    "adar2.df.cta <- subset(CTA_sites, ADAR==2)\n",
    "\n",
    "adar1.plt.cta <- ggplot(adar1.df.cta, aes(x=basenum, y=conversion_ratio)) + \n",
    "        geom_point(color='black', size=2, pch=21, fill='tan', alpha=0.6) + \n",
    "        facet_grid(row=vars(insert), col=vars(rnaseh)) +\n",
    "        labs(y='C / (C+T)', x='Read position', title='ADAR1 deaimination at CTA sites') + theme_pubclean()\n",
    "\n",
    "adar2.plt.cta <- ggplot(adar2.df.cta, aes(x=basenum, y=conversion_ratio)) + \n",
    "        geom_point(color='black', size=2, pch=21, fill='dodgerblue', alpha=0.6) + \n",
    "        facet_grid(row=vars(insert), col=vars(rnaseh)) +\n",
    "        labs(y='C / (C+T)', x='Read position', title='ADAR2 deaimination at CTA sites') + theme_pubclean()\n",
    "\n",
    "plt2 <- ggarrange(adar1.plt.cta, adar2.plt.cta)\n",
    "ggsave('plots/C_to_T_ratio_at_CTA_sites.png', plt2, dpi=300)\n",
    "plt2"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
